### PR TITLE
Add IVF changes to support Faiss byte vector

### PR DESCRIPTION
**Pull request by naveentatikonda**
### Description
Add IVF changes to support Faiss byte vector which behind the scenes uses Faiss SQ8_direct_signed scalar quantizer.

### Related Issues
https://github.com/opensearch-project/k-NN/issues/1659

### Check List
- [x] New functionality includes testing.
- [x] Commits are signed per the DCO using `--signoff`.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/knn/training/BinaryTrainingDataConsumer.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import org.opensearch.search.SearchHit;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Transfers binary vectors from JVM to native memory.
+ */
+public class BinaryTrainingDataConsumer extends TrainingDataConsumer {
+    private static final Logger logger = LogManager.getLogger(TrainingDataConsumer.class);
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/training/BinaryTrainingDataConsumer.java:22]: use @Log4j2

**Comment by navneet1v** [comment]
@naveentatikonda can we put a small benchmark run for this to showcase that ByteVector with IVF is working and we are seeing drop in memory usage.

File: jni/src/faiss_wrapper.cpp [hunk 2]
```diff
@@ -320,6 +320,96 @@ void knn_jni::faiss_wrapper::CreateBinaryIndexFromTemplate(knn_jni::JNIUtilInter
     faiss::write_index_binary(&idMap, indexPathCpp.c_str());
 }
 
+void knn_jni::faiss_wrapper::CreateByteIndexFromTemplate(knn_jni::JNIUtilInterface * jniUtil, JNIEnv * env, jintArray idsJ,
```
[the remark this answers is not in the document]
Comment (naveentatikonda) [reply jni/src/faiss_wrapper.cpp:323]: Yes, we need to refactor these methods and move to IndexService. But, including those refactoring changes into this PR will make it big so want to do it later. Junqiu already started refactoring train Indices in #1918. Will collab with Junqiu and add my changes to that PR

File: src/main/java/org/opensearch/knn/training/BinaryTrainingDataConsumer.java (hunk 1 above)
Comment (naveentatikonda) [reply src/main/java/org/opensearch/knn/training/BinaryTrainingDataConsumer.java:22]: Ack

**Comment by naveentatikonda** [comment]
> @naveentatikonda can we put a small benchmark run for this to showcase that ByteVector with IVF is working and we are seeing drop in memory usage.

This is the RSS metrics for Cohere-wiki-768 dataset (475,858 vectors) and the memory usage is as expected.
IVF memory requirement - 0.4 gb (appx.)
Heap - 1GB
Total memory provided - 1.7 GB

image

**Review by jmazanec15 — APPROVED** [review]
LGTM

File: src/main/java/org/opensearch/knn/index/util/IndexUtil.java [hunk 3]
```diff
@@ -170,20 +162,18 @@ public static ValidationException validateKnnField(
...
             if (trainRequestKnnMethodContext != null) {
                 MethodComponentContext methodComponentContext = trainRequestKnnMethodContext.getMethodComponentContext();
                 Map<String, Object> parameters = methodComponentContext.getParameters();
 
                 if (parameters != null && parameters.containsKey(KNNConstants.METHOD_ENCODER_PARAMETER)) {
                     MethodComponentContext encoder = (MethodComponentContext) parameters.get(KNNConstants.METHOD_ENCODER_PARAMETER);
-                    if (encoder != null
-                        && KNNConstants.ENCODER_PQ.equals(encoder.getName())
-                        && VectorDataType.BINARY == trainRequestVectorDataType) {
+                    if (encoder != null && VECTOR_DATA_TYPES_NOT_SUPPORTING_ENCODERS.contains(trainRequestVectorDataType)) {
```
Comment (junqiu-lei) [inline src/main/java/org/opensearch/knn/index/util/IndexUtil.java:172]: Thanks for the update. Btw, could you also check if the right datatype is validated when training?

Comment (naveentatikonda) [reply src/main/java/org/opensearch/knn/index/util/IndexUtil.java:172]: It is already taken care 
https://github.com/opensearch-project/k-NN/pull/2002/files#diff-572fd1fb31f6ec72d21051e6c123948cf758179a61fc6e50349f0be8123bd553R41

**Review by junqiu-lei — APPROVED** [review]
LGTM

**Pull request merged**